src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/AdvancedComposition.kt

```diff
@@ -0,0 +1,93 @@
+/**
+ * Copyright 2022 The Cross-Media Measurement Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ * ```
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * ```
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.wfanet.measurement.eventdataprovider.privacybudgetmanagement
+
+import kotlin.math.exp
+import kotlin.math.pow
+
+/** Memoized computation of binomial coefficients. */
+class Binom {
+  var memoizedCoeffs = HashMap<Pair<Int, Int>, Float>()
+
+  /**
+   * Computes the number of distinct ways to choose k items from a set of n.
+   *
+   * @param n The size of the set.
+   * @param k The size of the subset that is drawn.
+   * @return The number of distinct ways to draw k items from a set of size n. Alternatively, the
+   * coefficient of x^k in the expansion of (1 + x)^n.
+   */
+  fun coeff(n: Int, k: Int): Float {
+    if ((n < 0) || (k < 0) || (n < k)) {
+      return 0.0f
+    } else if ((k == 0) || (n == k)) {
+      return 1.0f
+    } else if (memoizedCoeffs.containsKey(Pair(n, k))) {
+      return memoizedCoeffs.get(Pair(n, k)) ?: 0.0f
```

Comment:
Optional: we can also let `k = n - k` in the case where k > n/2 to reduce the running time / memory further.